tiered-storage/jcloud/src/main/java/org/apache/bookkeeper/mledger/offload/jcloud/impl/BlobStoreManagedLedgerOffloader.java

```diff
@@ -230,6 +265,202 @@ public String getOffloadDriverName() {
+        final long blockLedgerId = peek.getLedgerId();
+        final long blockEntryId = peek.getEntryId();
+        payloadStream = new BufferedOffloadStream(streamingBlockSize, offloadBuffer, segmentInfo,
+                blockLedgerId, blockEntryId, bufferLength);
+        try {
+            streamingIndexBuilder.addLedgerMeta(blockLedgerId, ml.getClosedLedgerInfo(blockLedgerId).get());
+            streamingIndexBuilder.withDataBlockHeaderLength(StreamingDataBlockHeaderImpl.getDataStartOffset());
+        } catch (InterruptedException | ExecutionException e) {
+            offloadResult.completeExceptionally(e);
+            return;
+        }
+        log.debug("begin upload payload");
+        Payload partPayload = Payloads.newInputStreamPayload(payloadStream);
+        partPayload.getContentMetadata().setContentType("application/octet-stream");
+        streamingParts.add(blobStore.uploadMultipartPart(streamingMpu, partId, partPayload));
+        streamingIndexBuilder.addBlock(blockLedgerId, blockEntryId, partId, streamingBlockSize);
+
+        log.debug("UploadMultipartPart. container: {}, blobName: {}, partId: {}, mpu: {}",
+                config.getBucket(), streamingDataBlockKey, partId, streamingMpu.id());
+        if (segmentInfo.isClosed() && offloadBuffer.isEmpty()) {
+            log.debug("segment closed, buffer is empty");
+            try {
+                blobStore.completeMultipartUpload(streamingMpu, streamingParts);
+                streamingIndexBuilder.withDataObjectLength(dataObjectLength + streamingBlockSize);
+                final StreamingOffloadIndexBlock index = streamingIndexBuilder.buildStreaming();
+                final IndexInputStream indexStream = index.toStream();
+                final BlobBuilder indexBlobBuilder = blobStore.blobBuilder(streamingDataIndexKey);
+                DataBlockUtils.addVersionInfo(indexBlobBuilder, userMetadata);
+                final InputStreamPayload indexPayLoad = Payloads.newInputStreamPayload(indexStream);
+                indexPayLoad.getContentMetadata().setContentLength(indexStream.getStreamSize());
+                indexPayLoad.getContentMetadata().setContentType("application/octet-stream");
+                final Blob indexBlob = indexBlobBuilder.payload(indexPayLoad)
+                        .contentLength(indexStream.getStreamSize())
+                        .build();
+                blobStore.putBlob(config.getBucket(), indexBlob);
+
+                offloadResult.complete(segmentInfo.result());
+            } catch (Exception e) {
+                log.error("streaming offload failed", e);
+                offloadResult.completeExceptionally(e);
+            }
+            log.debug("offload done");
+        } else {
+            log.debug("continue offload loop");
+            scheduler.chooseThread(segmentInfo)
+                    .execute(() -> streamingOffloadLoop(partId + 1, dataObjectLength + streamingBlockSize));
```

Comment:
here will be a serious problem because the streaming offload will retry to get entry from the queue for every 100ms, if a topic does not write any messages, the thread will available for other topic offloading until the offload segment timeout(closed) right?

Reply:
I used schedule instead of sleep